File: EdTechClient/src/app/services/booking.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { Booking } from '../models';

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  private apiUrl = 'http://localhost:5237/api/bookings';

  constructor(private http: HttpClient) { }

  createBooking(booking: Booking): Observable<Booking> {
    return this.http.post<Booking>(this.apiUrl, booking).pipe(
      catchError(error => {
        console.error('Error creating booking:', error);
        return of(booking);
      })
    );
  }

  getBookingsByStudent(studentId: string): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/student/${studentId}`).pipe(
      catchError(error => {
        console.error('Error fetching student bookings:', error);
        return of([]);
      })
    );
  }

  getBookingsByTutor(tutorId: string): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/tutor/${tutorId}`).pipe(
      catchError(error => {
        console.error('Error fetching tutor bookings:', error);
        return of([]);
      })
    );
  }

  getBookingById(id: string): Observable<Booking | undefined> {
    return this.http.get<Booking>(`${this.apiUrl}/${id}`).pipe(
      catchError(error => {
        console.error('Error fetching booking:', error);
        return of(undefined);
      })
    );
  }

  updateBookingStatus(id: string, status: string): Observable<Booking | undefined> {
    return this.http.put<Booking>(`${this.apiUrl}/${id}/status`, { status }).pipe(
      catchError(error => {
        console.error('Error updating booking status:', error);
        return of(undefined);
      })
    );
  }

  cancelBooking(id: string): Observable<boolean> {
    return this.http.delete<any>(`${this.apiUrl}/${id}`).pipe(
      catchError(error => {
        console.error('Error cancelling booking:', error);
        return of(false);
      })
    );
  }

  getUpcomingBookings(studentId: string): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/student/${studentId}/upcoming`).pipe(
      catchError(error => {
        console.error('Error fetching upcoming bookings:', error);
        return of([]);
      })
    );
  }

  getCompletedBookings(studentId: string): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/student/${studentId}/completed`).pipe(
      catchError(error => {
        console.error('Error fetching completed bookings:', error);
        return of([]);
      })
    );
  }
}

File: EdTechAPIs/Services/BookingService.cs
using EdTechAPIs.Models;

namespace EdTechAPIs.Services;

public interface IBookingService
{
    Task<Booking> CreateBooking(Booking booking);
    Task<List<Booking>> GetBookingsByStudent(string studentId);
    Task<List<Booking>> GetBookingsByTutor(string tutorId);
    Task<Booking> GetBookingById(string id);
    Task<Booking> UpdateBookingStatus(string id, string status);
    Task<Booking> UpdateBookingPaymentStatus(string id, string paymentStatus);
    Task<bool> CancelBooking(string id);
    Task<List<Booking>> GetUpcomingBookings(string studentId);
    Task<List<Booking>> GetCompletedBookings(string studentId);
}

public class BookingService : IBookingService
{
    private List<Booking> _bookings = new();

    public async Task<Booking> CreateBooking(Booking booking)
    {
        _bookings.Add(booking);
        return await Task.FromResult(booking);
    }

    public async Task<List<Booking>> GetBookingsByStudent(string studentId)
    {
        return await Task.FromResult(_bookings.Where(b => b.StudentId == studentId).OrderByDescending(b => b.CreatedDate).ToList());
    }

    public async Task<List<Booking>> GetBookingsByTutor(string tutorId)
    {
        return await Task.FromResult(_bookings.Where(b => b.TutorId == tutorId).OrderByDescending(b => b.CreatedDate).ToList());
    }

    public async Task<Booking> GetBookingById(string id)
    {
        return await Task.FromResult(_bookings.FirstOrDefault(b => b.Id == id));
    }

    public async Task<Booking> UpdateBookingStatus(string id, string status)
    {
        var booking = _bookings.FirstOrDefault(b => b.Id == id);
        if (booking != null)
        {
            booking.Status = status;
        }
        return await Task.FromResult(booking);
    }

    public async Task<Booking> UpdateBookingPaymentStatus(string id, string paymentStatus)
    {
        var booking = _bookings.FirstOrDefault(b => b.Id == id);
        if (booking != null)
        {
            booking.PaymentStatus = paymentStatus;
        }
        return await Task.FromResult(booking);
    }

    public async Task<bool> CancelBooking(string id)
    {
        var booking = _bookings.FirstOrDefault(b => b.Id == id);
        if (booking == null) return await Task.FromResult(false);

        booking.Status = "cancelled";
        return await Task.FromResult(true);
    }

    public async Task<List<Booking>> GetUpcomingBookings(string studentId)
    {
        return await Task.FromResult(_bookings
            .Where(b =>
                b.StudentId == studentId &&
                b.SessionDate > DateTime.UtcNow &&
                b.Status != "cancelled"
            )
            .OrderBy(b => b.SessionDate)
            .ToList());
    }

    public async Task<List<Booking>> GetCompletedBookings(string studentId)
    {
        return await Task.FromResult(_bookings
            .Where(b =>
                b.StudentId == studentId &&
                b.Status == "completed"
            )
            .OrderByDescending(b => b.SessionDate)
            .ToList());
    }
}
